### ISSUE TITLE
Documents deleted by other clients when the PWA is closed do not get removed from cache

### ISSUE THREAD

**Issue by wildhart** [jamauro/offline#14]
The title says it all really: Documents deleted by other clients when the PWA is closed do not get removed from cache when the PWA is re-opened.

Steps to reproduce:

* Create a document in synced collection, and make sure it is synced to the PWA via a subscription.
* Force-close the PWA.
* In another client (or in `meteor mongo` during dev), delete the document.
* Open the PWA (online) → The document still exists.
* Calling a method from the PWA to delete the document doesn't appear to work because the document is not removed from the cache.
* The only way to remove the document is to call `clear()` on the collection, or `clearAll()`, or clear the site storage.

**Comment by jamauro** [comment]
Thanks for reporting. I've been playing around with some ideas. I think the cleanest way would be to use a soft delete mechanism. During the syncing process, this package can use the soft delete flag as part of its query when it's reconciling with the server. It'd probably make sense to make this soft delete flag name configurable – defaulting to `deleted`.

Curious if you have any ideas for other solutions.

**Comment by wildhart** [comment]
That would work, if we want to use soft deletes and are aware of the requirement to configure the soft delete flag name.

I don't fully understand the cause of the problem because I haven't done a deep-dive into your code to see how the offline caching works, but I'm guessing it's not as simple as storing a list of cached document _ids before the subscrition starts receiving data from the server, wait for the subscription to finish loading, then delete any cached documents which weren't received in the new subscription?

I see you have a `_keep` method which does some server reconciliation.  What if, as part of your `_keep` method, you sent a list of cached document _ids to the server (per collection), then do a query on the server and return to the client a list of _ids which weren't found, so you can remove them from the cache?

**Comment by jamauro** [comment]
I've updated the Readme to recommend using soft deletes with some instructions on how to configure. I also created a new package [jam:soft-delete](https://github.com/jamauro/soft-delete) to help make it easier for people who don't already have a soft-delete mechanism.

> I'm guessing it's not as simple as storing a list of cached document _ids before the subscrition starts receiving data from the server, wait for the subscription to finish loading, then delete any cached documents which weren't received in the new subscription

I prototyped this solution but it felt a little brittle and it gets pretty complicated with the different subscriptions to the point where it kind of made sense to restrict it to a "pick up where you left off" solution, effectively limiting the amount of data you stored offline. This might be a fine solution for some apps, but felt a little too restrictive. Maybe in the future if people really want this, it could be explored again. Or maybe someone has a better idea that I didn't consider.

> What if, as part of your _keep method, you sent a list of cached document _ids to the server (per collection), then do a query on the server and return to the client a list of _ids which weren't found, so you can remove them from the cache

I actually started with this initially and was going to ship it as the primary solution in the first version of this package but ultimately decided to ditch it and use a server-based timestamp to reconcile. It felt less than ideal to send a bunch of ids to the server every time the user connected.
